src/Handler/Workspace/Create/CreateWorkspaceHandler.php

```diff
@@ -0,0 +1,132 @@
+    public const DEFAULT_SPOOL_SPACE_SIZE = 1e9; // 1GB
+
+    private TeradataSessionManager $manager;
+
+    public function __construct(TeradataSessionManager $manager)
+    {
+        $this->manager = $manager;
+    }
+
+    /**
+     * @inheritDoc
+     * @param GenericBackendCredentials $credentials
+     * @param CreateWorkspaceCommand $command
+     */
+    public function __invoke(
+        Message $credentials,
+        Message $command,
+        array $features
+    ): ?Message {
+        assert($credentials instanceof GenericBackendCredentials);
+        assert($command instanceof CreateWorkspaceCommand);
+
+        // validate
+        assert(!empty($command->getStackPrefix()), 'CreateWorkspaceCommand.stackPrefix is required');
+        assert(!empty($command->getWorkspaceId()), 'CreateWorkspaceCommand.workspaceId is required');
+        assert(!empty($command->getProjectUserName()), 'CreateWorkspaceCommand.projectUserName is required');
+        assert(!empty($command->getProjectRoleName()), 'CreateWorkspaceCommand.projectRoleName is required');
+        assert(!empty($command->getProjectReadOnlyRoleName()), 'CreateWorkspaceCommand.projectReadOnlyRoleName is required');
+
+        $db = $this->manager->createSession($credentials);
+
+        // root user is also root database
+        $databaseName = $credentials->getPrincipal();
+
+        // allow override spaces for user
+        /** @var CreateWorkspaceCommand\CreateWorkspaceTeradataMeta|null $meta */
+        $meta = MetaHelper::getMetaRestricted($command, CreateWorkspaceCommand\CreateWorkspaceTeradataMeta::class);
+        $permSpace = self::DEFAULT_PERM_SPACE_SIZE;
+        $spoolSpace = self::DEFAULT_SPOOL_SPACE_SIZE;
+        if ($meta !== null) {
+            $permSpace = $meta->getPermSpace() !== '' ? $meta->getPermSpace() : $permSpace;
+            $spoolSpace = $meta->getSpoolSpace() !== '' ? $meta->getSpoolSpace() : $spoolSpace;
+        }
+
+        $nameGenerator = NameGeneratorFactory::getGeneratorForBackendAndPrefix(
+            BackendSupportsInterface::BACKEND_TERADATA,
+            $command->getStackPrefix()
+        );
+        $newWsUserName = $nameGenerator->createWorkspaceUserNameForWorkspaceId($command->getWorkspaceId());
+        $newWsRoleName = $nameGenerator->createWorkspaceRoleNameForWorkspaceId($command->getWorkspaceId());
+        $newWsPassword = Password::generate(
+            30,
+            Password::SET_LOWERCASE | Password::SET_UPPERCASE | Password::SET_NUMBER | Password::SET_SPECIAL_CHARACTERS
+        );
+
+        $db->executeStatement(sprintf(
+            'CREATE ROLE %s;',
+            TeradataQuote::quoteSingleIdentifier($newWsRoleName),
+        ));
+
+        // grant workspace role to root user
+        $db->executeStatement(sprintf(
+            'GRANT %s TO %s;',
+            TeradataQuote::quoteSingleIdentifier($newWsRoleName),
+            TeradataQuote::quoteSingleIdentifier($credentials->getPrincipal()),
+        ));
```

Comment:
přemýšlám jestli je to potřeba jestli ta role není zrovna delegovaná a taky tu WS budeš dělat pod projektovýma credentials asi takže děláš vlastně 2x stejný grant.

Reply:
tohle je blbost - nemam credentials root usera - tady ten `$credentials->getPrincipal()` je ve skutecnosti `projectUser` --> odstraneno